lib/rails_twirp/testing/integration_test.rb

```diff
@@ -17,25 +23,33 @@ def before_rpc(&block)
       @before_rpc << block
     end
 
-    def rpc(service, rpc, request, headers: nil)
+    def rpc(svc, rpc, request, headers: nil)
       @request = request
-      service = app.twirp.routes.services[service].to_service
+      input_class = svc.rpcs[rpc][:input_class]
+      output_class = svc.rpcs[rpc][:output_class]
 
-      rack_env = {
-        "HTTP_HOST" => "localhost"
+      content_type = Twirp::Encoding::PROTO
+      env = {
+        "CONTENT_TYPE" => content_type,
+        "HTTP_HOST" => "localhost",
+        "PATH_INFO" => "#{@mount_path}/#{svc.service_full_name}/#{rpc}",
+        "REQUEST_METHOD" => "POST"
       }
-      http_request = ActionDispatch::Request.new(rack_env)
+      http_request = ActionDispatch::Request.new(env)
       http_request.headers.merge! headers if headers.present?
-      env = {rack_env: rack_env}
+
+      env["rack.input"] = StringIO.new(Twirp::Encoding.encode(request, input_class, content_type))
 
       @before_rpc.each do |hook|
         hook.call(env)
       end
 
-      response = begin
-        service.call_rpc rpc, request, env
-      rescue => e
-        Twirp::Error.internal_with(e)
+      status, headers, body = app.call(env)
+      body = body.join
+      response = if status === 200
```

Comment:
Don't know if this would nicely extract into a helper, but maybe a comment to explain you're decoding the Twirp response